File: src/types/operations.ts
import { Value } from "./values";
import type { LiveObject, LiveMap, LiveList, LiveCounter } from "./values";
import type { BatchContext } from "./batchcontext";

// Options that can be provided to operation methods which specify
// additional data to include on the published object message(s).
export interface OperationOptions {
  id?: string;
  extras?: {
    headers?: Record<string, string>;
    [key: string]: any;
  };
  encoding?: string;
}

interface BatchOperation<T extends LiveObject> {
  // Batch multiple operations together using a batch context, which
  // wraps the underlying PathObject or Instance that batch was called from.
  // The batch context always contains a specific resolved instance, even
  // if called from a PathObject. If a specific instance cannot be obtained
  // from the referenced path, batch will throw an error.
  batch(fn: (ctx: BatchContext<T>) => void): Promise<void>;
}

export interface LiveMapOperations<
  T extends Record<string, Value> = Record<string, Value>,
> extends BatchOperation<LiveMap<T>> {
  set<K extends keyof T>(
    key: K,
    value: T[K],
    options?: OperationOptions,
  ): Promise<void>;
  remove(key: keyof T, options?: OperationOptions): Promise<void>;
}

export interface LiveListOperations<T extends Value = Value>
  extends BatchOperation<LiveList<T>> {
  append(value: T, options?: OperationOptions): Promise<void>;
  prepend(value: T, options?: OperationOptions): Promise<void>;
  pop(options?: OperationOptions): Promise<void>;
  shift(options?: OperationOptions): Promise<void>;
  insert(index: number, value: T, options?: OperationOptions): Promise<void>;
  remove(index: number, options?: OperationOptions): Promise<void>;
}

export interface LiveCounterOperations extends BatchOperation<LiveCounter> {
  increment(amount?: number, options?: OperationOptions): Promise<void>;
}

export interface AnyOperations {
  batch<T extends LiveObject = LiveObject>(
    fn: (ctx: BatchContext<T>) => void,
  ): Promise<void>;

  // LiveMap operations
  set<T extends Record<string, Value> = Record<string, Value>>(
    key: keyof T,
    value: T[keyof T],
    options?: OperationOptions,
  ): Promise<void>;
  remove<T extends Record<string, Value> = Record<string, Value>>(
    key: keyof T,
    options?: OperationOptions,
  ): Promise<void>;

  // LiveList operations
  append<T extends Value = Value>(
    value: T,
    options?: OperationOptions,
  ): Promise<void>;
  prepend<T extends Value = Value>(
    value: T,
    options?: OperationOptions,
  ): Promise<void>;
  pop(options?: OperationOptions): Promise<void>;
  shift(options?: OperationOptions): Promise<void>;
  insert<T extends Value = Value>(
    index: number,
    value: T,
    options?: OperationOptions,
  ): Promise<void>;
  remove<T extends Value = Value>(
    index: number,
    options?: OperationOptions,
  ): Promise<void>;

  // LiveCounter operations
  increment(amount?: number, options?: OperationOptions): Promise<void>;
}

File: swift/Sources/AblyLiveObjects/operations.swift
// Options that can be provided to operation methods which specify
// additional data to include on the published object message(s).
public struct OperationOptions {
    public init(id: String? = nil, extras: Extras? = nil, encoding: String? = nil) {
        self.id = id
        self.extras = extras
        self.encoding = encoding
    }

    var id: String?
    var extras: Extras?
    var encoding: String?
}

/*
  // Lawrence: skipped
 // Batch multiple operations together using a batch context, which
 // wraps the underlying PathObject or Instance that batch was called from.
 // The batch context always contains a specific resolved instance, even
 // if called from a PathObject. If a specific instance cannot be obtained
 // from the referenced path, batch will throw an error.
 protocol BatchOperation {
   func batch(fn: (BatchContext) -> Void) async throws
 }
  */

public protocol LiveMapOperations {
    func set(key: String, value: Value) async throws
    func set(key: String, value: Value, options: OperationOptions) async throws
    func remove(key: String) async throws
    func remove(key: String, options: OperationOptions) async throws
}

public protocol LiveListOperations {
    func append(_ value: Value) async throws
    func append(_ value: Value, options: OperationOptions) async throws
    func prepend(_ value: Value) async throws
    func prepend(_ value: Value, options: OperationOptions) async throws
    func pop() async throws
    func pop(options: OperationOptions) async throws
    func shift() async throws
    func shift(options: OperationOptions) async throws
    func insert(at index: Int, value: Value) async throws
    func insert(at index: Int, value: Value, options: OperationOptions) async throws
    func remove(at index: Int) async throws
    func remove(at index: Int, options: OperationOptions) async throws
}

public protocol LiveCounterOperations {
    func increment() async throws
    func increment(amount: Double) async throws
    func increment(options: OperationOptions) async throws
    func increment(amount: Double, options: OperationOptions) async throws
}

public protocol UnionOperations {
    // LiveMap operations
    func set(key: String, value: Value) async throws
    func set(key: String, value: Value, options: OperationOptions) async throws
    func remove(key: String) async throws
    func remove(key: String, options: OperationOptions) async throws

    // LiveList operations
    func append(_ value: Value) async throws
    func append(_ value: Value, options: OperationOptions) async throws
    func prepend(_ value: Value) async throws
    func prepend(_ value: Value, options: OperationOptions) async throws
    func pop() async throws
    func pop(options: OperationOptions) async throws
    func shift() async throws
    func shift(options: OperationOptions) async throws
    func insert(at index: Int, value: Value) async throws
    func insert(at index: Int, value: Value, options: OperationOptions) async throws
    func remove(at index: Int) async throws
    func remove(at index: Int, options: OperationOptions) async throws

    // LiveCounter operations
    func increment() async throws
    func increment(amount: Double) async throws
    func increment(options: OperationOptions) async throws
    func increment(amount: Double, options: OperationOptions) async throws
}
